packages/client/src/objectSet/ObjectSetListenerWebsocket.test.ts

```diff
@@ -0,0 +1,494 @@
+    addLoggerToApiServer(rootLogger);
+  });
+
+  afterAll(() => {
+    apiServer.close();
+  });
+
+  describe("basic setup", () => {
+    let minimalClient: MinimalClient;
+
+    const MINIMUM_RECONNECT_DELAY = 2_000;
+    const OBJECT_SET_EXPIRY_MS = 60_000;
+
+    let client: ObjectSetListenerWebsocket;
+    let listener: MockedObject<
+      Required<ObjectSetListener<ObjectOrInterfaceDefinition>>
+    >;
+    let oslwInst = 0;
+
+    beforeEach(() => {
+      minimalClient = createMinimalClient(
+        Ontology.metadata,
+        STACK,
+        async () => "myAccessToken",
+        { logger: rootLogger },
+      );
+
+      client = new ObjectSetListenerWebsocket({
+        ...minimalClient,
+        logger: rootLogger.child({ oslwInst: oslwInst++ }),
+      }, {
+        minimumReconnectDelayMs: MINIMUM_RECONNECT_DELAY,
+        objectSetExpiryMs: OBJECT_SET_EXPIRY_MS,
+      });
+
+      listener = {
+        onChange: vi.fn(),
+        onError: vi.fn(),
+        onOutOfDate: vi.fn(),
+      };
+
+      let objectSetRidCounter = 0;
+      apiServer.use(
+        http.post(
+          `${STACK}/object-set-service/api/objectSets/temporary`,
+          () =>
+            HttpResponse.json({
+              objectSetRid: `rid.hi.${objectSetRidCounter++}`,
+            }),
+        ),
+        http.post(
+          `${STACK}/object-set-watcher/api/object-set-watcher/batchEnableWatcher`,
+          async () => {
+            return HttpResponse.json({});
+          },
+        ),
+      );
+
+      vi.useFakeTimers();
+    });
+
+    afterEach(() => {
+      vi.restoreAllMocks();
+      apiServer.resetHandlers();
+    });
+
+    it("should not send onOutOfDate since it has not subscribed yet", () => {
+      expect(MockedWebSocket).not.toHaveBeenCalled();
+      expect(listener.onOutOfDate).not.toHaveBeenCalled();
+      expect(listener.onChange).not.toHaveBeenCalled();
+      expect(listener.onError).not.toHaveBeenCalled();
+    });
+
+    describe("requests subscription", () => {
+      let ws: MockedWebSocket;
+      let unsubscribe: () => void;
+      let subReq1: ObjectSetSubscribeRequests;
+
+      beforeEach(async () => {
+        [ws, unsubscribe] = await subscribeAndExpectWebSocket(
+          client,
+          listener,
+        );
+
+        subReq1 = await expectSingleSubscribeMessage(ws);
+      });
+
+      afterEach(() => {
+        // just ensure we ended in good states, as we clear these as we consume them
+        expect(MockedWebSocket).not.toHaveBeenCalled();
+
+        // make sure we are unsubscribed so closing the socket doesn't trigger reconnects
+        unsubscribe();
+
+        // disconnect the socket
+        setWebSocketState(ws, "close");
+
+        vi.runAllTicks();
+
+        // Make sure we cleaned up
+        expectEqualRemoveAndAddListeners(ws);
+      });
+
+      it("only sends a single request", async () => {
+        respondSuccessToSubscribe(ws, subReq1);
+        // actually this is broken FIXME
```

Comment:
What exactly is broken here?

Reply:
There are scenarios where you can get multiple send requests, I just didnt mock the timing of them all out.